### PR TITLE
Install `__tcfapiLocator` and post message listener

### PR DESCRIPTION
**Pull request by micajuine-ho**
Partial for #30385

Under Experiment flag `'tcf-post-message-proxy-api'`.

If the `exposesTcfApi: true` signal is detected, then we will install a top level event listener for PostMessages. 

Additionally, we will install an iframe named `__tcfapiLocator` within the `amp-consent` element (styled to be fully hidden). Once the 3p iframe is unblocked via amp-consent, it will be able to locate this iframe, which will signal to it that it can send post messages to the parent window that follow the [`__tcfapiCall`](https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20CMP%20API%20v2.md#how-can-vendors-that-use-iframes-call-the-cmp-api-from-an-iframe) structure.

The installed listener will listen for `__tcfapiCall` Post Messages, validate them (in terms of what AMP supports), and then check that the message came from an iframe attached to the page, before triggering the correct method (todo in the next PR).

Added unit tests as well as a manual DIY example utilizing `amp-video-iframe` as the extension that loads a 3p iframe.

Quick ReadMe: https://docs.google.com/document/d/13rgLi2M9ujvDZhk15UCU_ODd3KwXUTmHnhobXNEsTOY/edit?pli=1#heading=h.845z2apkjpop

EDIT: PENDING LEGAL REVIEW

### CODE REVIEW

File: extensions/amp-consent/0.1/amp-consent.js
Comment (zhouyx) [inline extensions/amp-consent/0.1/amp-consent.js:844]: in non AMP page, CMP creates the iframe so other scripts on the same page can locate the iframe and understand whether TCF API is supported. What's the point of creating the iframe in AMP page?

File: extensions/amp-consent/0.1/consent-info.js
Comment (zhouyx) [inline extensions/amp-consent/0.1/consent-info.js:465]: should the AMP provided APIs be optional and defined in `exposes` as well?

File: extensions/amp-consent/0.1/amp-consent.js
Comment (zhouyx) [inline extensions/amp-consent/0.1/amp-consent.js:575]: what happens if a message is received beforehand?

File: extensions/amp-consent/0.1/consent-info.js
Comment (zhouyx) [inline extensions/amp-consent/0.1/consent-info.js:78]: Will we expand the list later on?

**Review by zhouyx — CHANGES_REQUESTED** [review]
I have a few questions on 
1. the performance impact to support the API
2. Future support if the TCF API changes
3. A complete list of what's supported, what's missing in AMP

Given this is not a trivial change. I recommend getting the design reviewed more broadly and also have @nainar agreement on this support.

**Comment by micajuine-ho** [comment]
Thanks for these questions!

> 1. the performance impact to support the API

[Described in the design doc](https://docs.google.com/document/d/13rgLi2M9ujvDZhk15UCU_ODd3KwXUTmHnhobXNEsTOY/edit#heading=h.6wcwtpyoumbv), but it will have very little performance impact, we're just inserting an empty iframe onto the doc (no scripts running on it), as well as adding an event listener to send/receive post messages to 3p iframes (after they've been unblocked).

> 2. Future support if the TCF API changes

As this feature is meant to make vendors and publishers lives easier, we will try best-effort (upon request) to update or create a new feature version to support the changes. i.e. if TCF V3 postMessage API is created, then we can add `tcfPostMessageApi3.0` to our list of exposed APIs.

WDYT?

> 3. A complete list of what's supported, what's missing in AMP

Missing support for: 
- the [__tcfApiLocator API](https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20CMP%20API%20v2.md#details-for-vendors) (where 3P iframe searches parent windows for `__tcfApiLocator` iframe)
- [TCF 2.0 PostMessage API](https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20CMP%20API%20v2.md#how-does-the-cmp-provide-the-api) (where 3P iframes will send post messages to the correct parent window and have consent information returned in the form of TCData to determine what content to show in the iframe).

This is solving the problem of letting 3P iframes that are created through amp extensions have the ability to reuse existing non-AMP code, to gather TCData. These APIs act as a proxy between the integrated CMP and the vendors, where normally the CMP would provide these APIs on the publishers page.

> Given this is not a trivial change. I recommend getting the design reviewed more broadly and also have @nainar agreement on this support.

We got approval from @jeffjose  before he went on leave--but I will loop @nainar in. I will take this to design review as well.

File: extensions/amp-consent/0.1/amp-consent.js
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/amp-consent.js:844]: discussed offline, but the iframe just acts as a flag that the PostMessasge API is supported in the **parent** window (https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/TCFv2/IAB%20Tech%20Lab%20-%20CMP%20API%20v2.md#is-there-a-sample-iframe-script-call-to-the-cmp-api)

File: extensions/amp-consent/0.1/consent-info.js
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/consent-info.js:78]: We could in the future. Do you suggest we just change it to a string for now, and then if we do expand later on, make the change then?

File: extensions/amp-consent/0.1/amp-consent.js
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/amp-consent.js:575]: Ahh yes, that's a good point. After our convo today, I think we don't have to allow CMP to send this signal via `checkConsentHref`, instead they can just have it in their checked in config or request that pubs add it themselves. WDYT?

File: extensions/amp-consent/0.1/consent-info.js
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/consent-info.js:465]: Actually for now, since we are thinking about performance, I'm ok with turning this into `exposesTcfApi: true` instead of `exposes: []`. This will mean less overall code and we can easily expand in the future. WDYT?

File: extensions/amp-consent/0.1/amp-consent.js
Comment (zhouyx) [reply extensions/amp-consent/0.1/amp-consent.js:575]: sounds reasonable to me. Whether to support the API or not should be a static decision, and there's no need to get it remotely.

File: extensions/amp-consent/0.1/consent-info.js
Comment (zhouyx) [reply extensions/amp-consent/0.1/consent-info.js:465]: My hope is that we won't expand this in the future. If we have it, it should be more a scalable solution. That the API should not be implemented by the AMP team, maybe in the format of a plugin.

**Comment by zhouyx** [comment]
Add @jridgewell as reviewer, who is more knowledgeable on the cost of iframes.

File: extensions/amp-consent/0.1/amp-consent.js
Comment (zhouyx) [inline extensions/amp-consent/0.1/amp-consent.js:778]: In theory, it's still possible that the message arrives before the event listener is registered. Or the 3p code start to looking for `TCF_API_LOCATOR` before the iframe is created. What does the TCF API suggest in this scenario?

File: extensions/amp-consent/0.1/amp-consent.js
Comment (zhouyx) [inline extensions/amp-consent/0.1/amp-consent.js:819]: Before we actually implement the event handler, should we wrap the function in experiment? Or should we throw a user error at this point?

File: extensions/amp-consent/0.1/amp-consent.js
Comment (jridgewell) [inline extensions/amp-consent/0.1/amp-consent.js:776]: Can amp-consent be inside an FIE? Will we need to check the top-window if it is?

File: extensions/amp-consent/0.1/amp-consent.js
Comment (jridgewell) [reply extensions/amp-consent/0.1/amp-consent.js:778]: When is the 3p setup?

**Review by jridgewell — APPROVED** [review]
Owners approval

File: extensions/amp-consent/0.1/amp-consent.js
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/amp-consent.js:819]: Let's do an experiment.

File: extensions/amp-consent/0.1/amp-consent.js
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/amp-consent.js:778]: > In theory, it's still possible that the message arrives before the event listener is registered. Or the 3p code start to looking for TCF_API_LOCATOR before the iframe is created. What does the TCF API suggest in this scenario?

There is no recommendation for what happens if the parent doc event listener is not registered before the message arrives. 

If the `TCF_API_LOCATOR` window is not present when the iframe loads, it is assumed that no CMP is present (and so the tcfApi is not supported). 

In theory, both of these could happen but only if the publisher does not tag the element with `data-block-on-consent`. For this API we state that elements that create iframes that could use this API need to be blocked with `data-block-on-consent` (as they already should be if they want to collect consent information anyway). Additionally, CMPs working with publishers should tell them to include this `exposesTcfApi` and also block certain elements with `data-block-on-consent`.

> When is the 3p setup?

3p setup occurs once the other extensions (amp-iframe, amp-video-iframe etc) loads the iframe. We state that for this API to work, these components need to be blocked by `data-block-on-consent` (to ensure that the `TCF_API_LOCATOR ` window is already created).

File: extensions/amp-consent/0.1/amp-consent.js
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/amp-consent.js:780]: amp-consent show not be in FIE in general, it also doesn't make sense for an ad to be using this API

**Review by zhouyx — CHANGES_REQUESTED** [review]
Please declare the experiment before using

**Comment by micajuine-ho** [comment]
> Please also declare the experiment

Ugh, bad merge. Thanks for the eyes :)

**Review by zhouyx — APPROVED** [review]
Please declare the experiment in `experiments-config.js`.
LGTM

File: examples/amp-consent/diy-3p-iframe-tcf-postmessage.html
Comment (dmanek) [inline examples/amp-consent/diy-3p-iframe-tcf-postmessage.html:144]: nit: fix formatting

File: extensions/amp-consent/0.1/test/test-amp-consent.js
Comment (dmanek) [inline extensions/amp-consent/0.1/test/test-amp-consent.js:853]: Wouldn't toggling the experiments be better in `beforeAll` and `afterAll`?

**Review by dmanek — APPROVED** [review]

File: extensions/amp-consent/0.1/test/test-amp-consent.js
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/test/test-amp-consent.js:853]: I'm not sure we have that functionality built in?

File: extensions/amp-consent/0.1/tcf-api-commands.js
Comment (dmanek) [inline extensions/amp-consent/0.1/tcf-api-commands.js:26]: nit: change the param to command (See Principle of Least Knowledge)

**Review by dmanek — APPROVED** [review]

File: extensions/amp-consent/0.1/tcf-api-commands.js
Comment (micajuine-ho) [reply extensions/amp-consent/0.1/tcf-api-commands.js:26]: We will need the full payload in the next PR.

**Pull request merged**
